cuML/src/common/cumlHandle.hpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cublas_v2.h>
+#include <cusolverDn.h>
+#include <cusparse.h>
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+class cumlHandle_impl {
+public:
+    cumlHandle_impl();
+    ~cumlHandle_impl();
+    void setStream( cudaStream_t stream );
+    cudaStream_t getStream() const;
+    void setDeviceAllocator( std::shared_ptr<deviceAllocator> allocator );
+    std::shared_ptr<deviceAllocator> getDeviceAllocator() const;
+    void setHostAllocator( std::shared_ptr<hostAllocator> allocator );
+    std::shared_ptr<hostAllocator> getHostAllocator() const;
+
+    cublasHandle_t getCublasHandle( int dev_idx = 0 ) const;
+    cusolverDnHandle_t getcusolverDnHandle( int dev_idx = 0 ) const;
+    cusparseHandle_t getcusparseHandle( int dev_idx ) const;
+
+    void waitOnUserStream() const;
+    void waitOnInternalStreams() const;
+
+    void setDevice( int dev_id );
+    void setDevices( const std::vector<int>& dev_ids );
+    int  getDevice( int dev_idx = 0 ) const;
+    int  getNumDevices() const;
+
+    cudaStream_t getDeviceStream( int dev_idx ) const;
+private:
+    std::vector<int>                    _dev_ids;
+    std::vector<cudaStream_t>           _streams;
+    std::vector<cublasHandle_t>         _cublas_handles;
+    std::vector<cusolverDnHandle_t>     _cusolverDn_handles;
+    std::vector<cusparseHandle_t>       _cusparse_handles;
+    std::shared_ptr<deviceAllocator>    _deviceAllocator;
```

Comment:
I'm trying to understand the need for the allocators to be `shared_ptr`s instead of a `unique_ptr` member.

Shouldn't the `handle` be in charge of maintaining the lifetime of the allocator? That is, who else would have a reference to the allocator that requires using a `shared_ptr`?

Reply:
I have chosen a shared pointer as this allows the cuML user to reuse an allocator in other places in a save way. Does that make sense?

Reply:
Yeah, I get that, but couldn't you also accomplish the same safety with a `unique_ptr`? 

My default is to always prefer a `unique_ptr` to a `shared_ptr` unless I absolutely need a `shared_ptr`, and I don't understand well enough yet why a `shared_ptr` is required here. 

My understanding is that there's a one-to-one relationship between an instance of a `ML::handle` and an instance of an allocator, and as such, the `handle` should have sole ownership over it's corresponding allocator. Is that understanding correct?

Reply:
From the point of view of anything that is implemented in cuML yes. However not from the point of view from a cuML user. If `setDeviceAllocator` would take a unique_ptr a cuML can't use a deviceAllocator that needs to outlive the cuML handle. Makes sense?

Reply:
> If `setDeviceAllocator` would take a unique_ptr a cuML can't use a deviceAllocator that needs to outlive the cuML handle. Makes sense? Or am I missing something?

Ah yes, so if someone wants to reuse the same allocator between two cuML API calls, using a `unique_ptr` would be difficult.

Thanks for explaining it to me, I believe a `shared_ptr` is absolutely appropriate here.